src/samplebuffer.h

```diff
@@ -0,0 +1,90 @@
+#ifndef SAMPLEBUFFER_H
+#define SAMPLEBUFFER_H
+
+#include "util/types.h" // CSAMPLE
+
+#include <algorithm> // std::swap
+
+// A sample buffer with properly aligned memory to enable SSE optimizations.
+//
+// No resize operation is provided intentionally for maximum efficiency!
+// If the size of an existing sample buffer needs to be altered after
+// construction this can simply be achieved by swapping the contents with
+// a temporary sample buffer that has been constructed with the appropriate
+// size:
+//
+//     SampleBuffer sampleBuffer(oldSize);
+//     ...
+//     SampleBuffer(newSize).swap(sampleBuffer);
+//     // Now the data in sampleBuffer is uninitialized and
+//     // sampleBuffer.size() == newSize
+//
+// No sample data is copied when swapping the contents of two samples buffers
+// for resizing!
+class SampleBuffer {
+    Q_DISABLE_COPY(SampleBuffer);
+
+  public:
+    typedef size_t size_type;
+    typedef CSAMPLE value_type;
```

Comment:
Ok, I stripped down SampleBuffer as much as I think is possible and sensible. Did you notice how I changed "size_type" from "size_t" to "int"? Only a single line of code was affected :)

Counter question: Why didn't you use a typedef if 'int' is essential for those loops to get optimized? This would be the perfect place in the code to document that decision. Moreover it would be helpful to keep the optimized loop code portable: The required type may vary between OSes (Windows, Linux, OS X), CPU architectures (32- vs 64-bit), compilers (GCC vs Clang), and even between different versions of the same compiler.

Need a simple example of what happens if you advertise native types in an API instead of typedefs? Have a look at __M4AHACK__ and you will see what I'm talking about ;)